JavaScript/JavaScriptSDK/ajax/ajax.ts

```diff
@@ -177,27 +178,24 @@ module Microsoft.ApplicationInsights {
             });
         }
 
-        private onReadyStateChangePrefix(xhr: XMLHttpRequestInstrumented) {
-            switch (xhr.readyState) {
-                case 3:
-                    xhr.ajaxData.responseStartedTime = dateTime.Now();
-                    break;
-                case 4:
-                    this.collectResponseData(xhr);
-                    break;
-            }
-        }
-
-        private onReadyStateChangePostfix(xhr: XMLHttpRequestInstrumented) {
-            if (xhr.readyState === 4) {
-                this.onAjaxComplete(xhr);
-            }
-        }
-
         private onAjaxComplete(xhr: XMLHttpRequestInstrumented) {
-            try {
+            xhr.ajaxData.responseFinishedTime = dateTime.Now();
+            xhr.ajaxData.status = xhr.status;
                 xhr.ajaxData.CalculateMetrics();
 
+            if (xhr.ajaxData.ajaxTotalDuration < 0) {
```

Comment:
Some time we send negative durations (based on investigation). This fix prevents it (by not sending negative ajaxes). This fix also adds extra telemetry (trace) which should help us understand why we get negative ajaxes.